# Week3a: Metabarcoding Journal Club

Before class, please read the following two journal articles in full:

Creer S, Deiner K, Frey S, Porazinska D, Taberlet P, Thomas WK, Potter C, Bik HM. (2016) The ecologist's field guide to sequence‐based identification of biodiversity. Methods in ecology and evolution, 7(9):1008-18.
http://onlinelibrary.wiley.com/doi/10.1111/2041-210X.12574/full

Keck F, Couton M, Altermatt F. (2023) Navigating the seven challenges of taxonomic reference databases in metabarcoding analyses. Molecular Ecology Resources, 23(4):742-55.
https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13746

-------------
#### 11:10am - 11:15am: 

* Fill out Minute cards: https://forms.gle/fK2FGG1uUSoaZTSo6
* Review course grading structure (in syllabus): https://www.dropbox.com/scl/fi/ty2m7xg97kvn285q97dxx/MARS-8180-EnvironmentalGenomics-syllabus-FINAL.pdf?rlkey=4mofi3z9awwkfk6yhbhoek1qs&dl=0 

#### 11:15am - 11:25am: What is Enviromental DNA?

<img width="796" alt="Screenshot 2025-01-20 at 6 04 28 PM" src="https://github.com/user-attachments/assets/bef11400-a085-4386-a6a1-2e684da1e02d" />

<img width="689" alt="Screenshot 2025-01-20 at 6 07 51 PM" src="https://github.com/user-attachments/assets/d85f7140-95f4-40f3-9a44-a5b834c48e54" />

#### 11:25-11:35am - In pairs, discuss the different sources of environmental DNA that may be present in your class dataset (e.g. the data you'll be working on, or another example marine dataset you're familiar with such as TARA Oceans, the Earth Microbiome Project,etc.)
* What is the main target source of eDNA/eRNA for your study? (the biological community or DNA/RNA pool you are aiming to analyze)
* What is the most important source of non-target eDNA/eRNA in your study? How will you quantify, characterize, and assess non-target eDNA/eRNA contributions during analysis?
* Are there any other sources of eDNA/eRNA that are important and pertinent to your study?

Follow GDoc questions (and take notes here) for this excercise: https://docs.google.com/document/d/19GQTbSbeNgVEeN9_9UlSsBTcVd33fpdGNF0fC6R_u6w/edit?usp=sharing

#### 11:35-11:45am - Groups report out from exercise + class discussion

#### 11:45 - 12:00pm - Typical -Omics Workflow, from sampling to data viz

<img width="1043" alt="Screenshot 2025-01-20 at 6 31 09 PM" src="https://github.com/user-attachments/assets/4429d03c-c91a-4c34-96b8-0269abd58bba" />

<img width="822" alt="Screenshot 2025-01-21 at 9 45 19 AM" src="https://github.com/user-attachments/assets/2c0dfec9-e216-432b-9d86-d52388d49e2b" />

<img width="888" alt="Screenshot 2025-01-21 at 9 45 49 AM" src="https://github.com/user-attachments/assets/2e1583cd-5ff8-4778-b533-ef50f191b706" />

#### 12:00-12:10pm - In pairs, discuss the different biases that your field collections and wet lab protocols may introduce in your class dataset.

* What is the most important source of bias that your fieldwork or sample collection strategies introduce into your dataset? (the biological community or DNA/RNA pool you are aiming to analyze)
* What is the most important source of bias that your wet lab protocols and sequencing strategy introduce into your dataset? (e.g. DNA extractions, PCR, library prep, the actual sequencing platform itself)
* What steps might you take to help you assess (and/or mitigate) the sources of bias you identified in your fieldwork/sampling? Wet lab and sequencing protocols?

Follow GDoc questions (and take notes here) for this excercise: https://docs.google.com/document/d/19GQTbSbeNgVEeN9_9UlSsBTcVd33fpdGNF0fC6R_u6w/edit?usp=sharing

#### 12:10-12:20pm - Groups report out from exercise + class discussion

#### 12:20-12:25pm - Wrap up and preview of next class
